bookstore/archive.py

```diff
@@ -29,6 +45,46 @@ def __init__(self, *args, **kwargs):
             s3_additional_kwargs={},
         )
 
+        self._thread_pool = ThreadPoolExecutor(max_workers=30)
+
+        self.locks: Dict[str, Lock] = {}
+
+    async def archive(self, record: ArchiveRecord):
+        """Process a storage write, only allowing one write at a time to each path
+        """
+        lock = self.locks.get(record.filepath)
+
+        try:
+            if lock is None:
+                lock = Lock()
+                self.locks[record.filepath] = lock
```

Comment:
I admit that I haven't used tornado or their locks, but this pattern seems broken.  If two threads try to lock a new filepath, they could both end up at line 59 at the same time.  Both will assign the lock at the same time (well sort of, last one would set the lock, but both threads would still have their own), and then you will have a race condition on write.  It isn't necessarily that bad, but I think that if we were being correct there should be a global lock that is required to create a new lock:

```
 if lock is None:
     async with self.locks.get('global'):
          lock = self.locks.get(record.filepath)
          if lock is None:
              lock = Lock()
              self.locks[record.filepath] = lock
```

Although I would probably just init that lock at the object level as that is a little too magical/coupled for my taste.

Reply:
Probably worth noting -- we're using threads here for one specific action that would block the event loop, which is writing to S3. We can't use the `Lock`s or `Queue`s in any of that threaded code.

From [tornado docs on locks](https://www.tornadoweb.org/en/stable/locks.html):

> Note that these primitives are not actually thread-safe and cannot be used in place of those from the standard library–they are meant to coordinate Tornado coroutines in a single-threaded app, not to protect shared objects in a multithreaded app.